File: webhooks/nodejs/samples/folder.created.js
// Create Subfolders for "HR Onboarding"

const express = require("express");
const crypto = require("crypto");
const axios = require("axios");
require("dotenv").config();

const app = express();
app.use(express.json({ verify: rawBodySaver }));

// Save raw body for signature verification
function rawBodySaver(req, res, buf) {
  req.rawBody = buf.toString();
}

// Generate HMAC-SHA256 signature
function getSecretHash(secret, payload) {
  return (
    "sha256=" +
    crypto.createHmac("sha256", secret).update(payload).digest("hex")
  );
}

app.post("/webhook", async (req, res) => {
  const signature = req.headers["x-docspace-signature-256"];
  const computedSignature = getSecretHash(process.env.WEBHOOK_SECRET_KEY, req.rawBody);

  if (signature !== computedSignature) {
    console.warn("Invalid signature");
    return res.sendStatus(401);
  }

  const { trigger } = req.body.event;
  const { id: folderId, title } = req.body.payload;

  // Triggered when a folder named "HR Onboarding" is created
  if (trigger === "folder.created" && title === "HR Onboarding") {
    console.log(`HR Onboarding folder created, creating subfolders...`);

    const subfolders = ["Documents", "Passport Photos", "Signed Forms"];

    try {
      for (const subfolderName of subfolders) {
        await axios.post(
          `${process.env.PORTAL_URL}/api/2.0/files/folder/${folderId}`,
          { title: subfolderName },
          {
            headers: {
              Authorization: `Bearer ${process.env.API_KEY}`,
              "Content-Type": "application/json"
            }
          }
        );

        console.log(`Subfolder created: ${subfolderName}`);
      }

      return res.sendStatus(200);
    } catch (err) {
      console.error("Error while creating HR subfolders:", err.response?.data || err.message);
      return res.sendStatus(500);
    }
  }

  res.sendStatus(200);
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Webhook server listening on port ${PORT}`);
});

File: webhooks/python/samples/folder.created.py
# Create Subfolders for "HR Onboarding"

from flask import Flask, request, abort
import hmac
import hashlib
import requests
import os

app = Flask(__name__)

# Generate HMAC-SHA256 signature
def get_secret_hash(secret, payload):
    return "sha256=" + hmac.new(secret.encode(), payload, hashlib.sha256).hexdigest()

@app.route("/webhook", methods=["POST"])
def webhook():
    signature = request.headers.get("x-docspace-signature-256")
    computed_signature = get_secret_hash(
        os.environ.get("WEBHOOK_SECRET_KEY"),
        request.get_data()
    )

    if signature != computed_signature:
        app.logger.warning("Invalid signature")
        abort(401)

    event = request.json.get("event", {})
    payload = request.json.get("payload", {})

    trigger = event.get("trigger")
    folder_id = payload.get("id")
    title = payload.get("title")

    # Triggered when a folder named "HR Onboarding" is created
    if trigger == "folder.created" and title == "HR Onboarding":
        app.logger.info("HR Onboarding folder created, creating subfolders...")

        subfolders = ["Documents", "Passport Photos", "Signed Forms"]

        try:
            for subfolder_name in subfolders:
                requests.post(
                    f"{os.environ.get('PORTAL_URL')}/api/2.0/files/folder/{folder_id}",
                    json={"title": subfolder_name},
                    headers={
                        "Authorization": f"Bearer {os.environ.get('API_KEY')}",
                        "Content-Type": "application/json"
                    }
                )

                app.logger.info(f"Subfolder created: {subfolder_name}")

            return "", 200

        except Exception as err:
            app.logger.error(f"Error while creating HR subfolders: {err}")
            return "", 500

    return "", 200

if __name__ == "__main__":
    port = int(os.environ.get("PORT", 3000))
    app.run(host="0.0.0.0", port=port)
